File: Linked List/leetcode_82.py
# Given the head of a sorted linked list, delete all nodes that have duplicate numbers, 
# leaving only distinct numbers from the original list. Return the linked list sorted as well.

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        # Dummy node to handle potential head deletion
        dummy = ListNode(0, head)
        prev = dummy
        curr = head

        while curr:
            while curr.next and curr.val == curr.next.val:
                curr = curr.next  # Skip duplicates

            if prev.next == curr:  # No duplicates found
                prev = prev.next
            else:
                prev.next = curr.next  # Remove duplicates

            curr = curr.next  # Move to the next node

        return dummy.next  # Return the modified list


# Example usage
head = ListNode(1, ListNode(2, ListNode(2, ListNode(3, ListNode(3)))))
solution = Solution()
new_head = solution.deleteDuplicates(head)

# Print the modified list
while new_head:
    print(new_head.val, end=" ")
    new_head = new_head.next

print()

File: Linked List/leetcode_82.cpp
// Problem Link ====>>https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/description/

#include<bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution
{
public:
    ListNode *deleteDuplicates(ListNode *head)
    {
        ListNode *dummy = new ListNode(0); // Dummy node to handle potential head deletion
        dummy->next = head;

        ListNode *prev = dummy;
        ListNode *current = head;

        while (current != nullptr)
        {
            while (current->next != nullptr && current->val == current->next->val)
            {
                current = current->next; // Skip duplicates
            }

            if (prev->next == current)
            { // No duplicates found
                prev = prev->next;
            }
            else
            {
                prev->next = current->next; // Remove duplicates
            }

            current = current->next; // Move to the next node
        }

        return dummy->next; // Return the modified list
    }
};

int main()
{
    // Create a sample linked list with duplicates
    ListNode *head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(2);
    head->next->next->next = new ListNode(3);
    head->next->next->next->next = new ListNode(3);

    Solution solution;                                   // Create an instance of the Solution class
    ListNode *newHead = solution.deleteDuplicates(head); // Call the function to remove duplicates

    // Print the modified list
    while (newHead != nullptr)
    {
        cout << newHead->val << " ";
        newHead = newHead->next;
    }

    cout << endl;

    return 0;
}
